### ISSUE TITLE
Question about enhanced decoding

### ISSUE THREAD

**Issue by mtybadger** [staoxiao/RetroMAE#19]
Hi staoxiao,

I wanted to ask more about how the enhanced decoding works - it looks like it generates 256 random possible attention masks, and then picks randomly from that list for each token. The label is just the original string with [CLS] and [SEP] masked for loss purposes. 

It looks like the input that the decoder layer gets is the CLS output of the encoder concatenated with the rest of the original string. What I'm confused about is that some of the masks the decoder tokens get have a 0 in the first position, so does this mean that not all the tokens in the decoder layer actually get to see the CLS token at all? This surely means they can't use the information in it to reconstruct the original text?

I hope my question makes sense, and I appreciate if there's something weird going on here with the attention that I don't understand - I don't see these naked QKV attention layers very often.

Spruce

**Comment by staoxiao** [comment]
Hi, thanks for your interest in our work!
We use the _whole_word_mask function to mask tokens, which will not mask the CLS token. You can refer to
https://github.com/huggingface/transformers/blob/v4.34.1/src/transformers/data/data_collator.py#L845.

**Comment by IT-coach-666** [comment]
> Hi, thanks for your interest in our work! We use the _whole_word_mask function to mask tokens, which will not mask the CLS token. You can refer to https://github.com/huggingface/transformers/blob/v4.34.1/src/transformers/data/data_collator.py#L845.


Hello, what you do is a wonderful work!   
And here I have a suggestion to imporve your pre-training process:
before use the _whole_word_mask, I think it should use ``random.seed(None)``, because your code use the 
``set_seed(42)`` by default, which will also set ``random.seed(42)`` also, so if not do what I have mentioned above, it will produce the same mask array (0 1 array) for the same text,  even thought you run many epoch, your data will not learn many mask position.

if what I mentioned above make sense? 

 I am a chinese coder from shanghai, and I run a Knowledge Base:
www.yuque.com/it-coach   
hope to make it better！

**Comment by staoxiao** [comment]
> > Hi, thanks for your interest in our work! We use the _whole_word_mask function to mask tokens, which will not mask the CLS token. You can refer to https://github.com/huggingface/transformers/blob/v4.34.1/src/transformers/data/data_collator.py#L845.
> 
> Hello, what you do is a wonderful work! And here I have a suggestion to imporve your pre-training process: before use the _whole_word_mask, I think it should use `random.seed(None)`, because your code use the `set_seed(42)` by default, which will also set `random.seed(42)` also, so if not do what I have mentioned above, it will produce the same mask array (0 1 array) for the same text, even thought you run many epoch, your data will not learn many mask position.
> 
> if what I mentioned above make sense?
> 
> I am a chinese coder from shanghai, and I run a Knowledge Base: [www.yuque.com/it-coach](http://www.yuque.com/it-coach) hope to make it better！

Hi, thanks for your advice! 
However, this command `random.seed(42)` will executed only once in the training process, so the mask for the same text will be different. I think there is no need to reset `random.seed(None)`. If you find the mask for the same text is the same, please tell me and I will analyze the reason.

**Comment by IT-coach-666** [comment]
> > > Hi, thanks for your interest in our work! We use the _whole_word_mask function to mask tokens, which will not mask the CLS token. You can refer to https://github.com/huggingface/transformers/blob/v4.34.1/src/transformers/data/data_collator.py#L845.
> > 
> > 
> > Hello, what you do is a wonderful work! And here I have a suggestion to imporve your pre-training process: before use the _whole_word_mask, I think it should use `random.seed(None)`, because your code use the `set_seed(42)` by default, which will also set `random.seed(42)` also, so if not do what I have mentioned above, it will produce the same mask array (0 1 array) for the same text, even thought you run many epoch, your data will not learn many mask position.
> > if what I mentioned above make sense?
> > I am a chinese coder from shanghai, and I run a Knowledge Base: [www.yuque.com/it-coach](http://www.yuque.com/it-coach) hope to make it better！
> 
> Hi, thanks for your advice! However, this command `random.seed(42)` will executed only once in the training process, so the mask for the same text will be different. I think there is no need to reset `random.seed(None)`. If you find the mask for the same text is the same, please tell me and I will analyze the reason.

oh, sorry  I mistake it.

**Comment by mtybadger** [comment]
Actually, while you're here, a question about DupMAE. It looks like the modeling_duplex file you included doesn't have the actual pooling operation you describe in the paper - how is the dimension reduction from (768, 768) performed? Is it something like a simple PCA or is it a trainable layer in itself?

Also, what do you think the impact would be of a concatenated vector for the embeddings in terms of the continuity of the vector space along different axes? Would the space be less interpretable because (if I'm following) the CLS token is "meant" to store semantic information while the aggregated OT embeddings store fine-grained word information, and they're being concatenated into one?

**Comment by IT-coach-666** [comment]
> > > > Hi, thanks for your interest in our work! We use the _whole_word_mask function to mask tokens, which will not mask the CLS token. You can refer to https://github.com/huggingface/transformers/blob/v4.34.1/src/transformers/data/data_collator.py#L845.
> > > 
> > > 
> > > Hello, what you do is a wonderful work! And here I have a suggestion to imporve your pre-training process: before use the _whole_word_mask, I think it should use `random.seed(None)`, because your code use the `set_seed(42)` by default, which will also set `random.seed(42)` also, so if not do what I have mentioned above, it will produce the same mask array (0 1 array) for the same text, even thought you run many epoch, your data will not learn many mask position.
> > > if what I mentioned above make sense?
> > > I am a chinese coder from shanghai, and I run a Knowledge Base: [www.yuque.com/it-coach](http://www.yuque.com/it-coach) hope to make it better！
> > 
> > 
> > Hi, thanks for your advice! However, this command `random.seed(42)` will executed only once in the training process, so the mask for the same text will be different. I think there is no need to reset `random.seed(None)`. If you find the mask for the same text is the same, please tell me and I will analyze the reason.
> 
> oh, sorry I mistake it.

yesterday, i mistake some features of the random.seed(), below is the example:

`
import random

# jy: 使得 seed 失效
#random.seed(None)
random.seed(42)


input_tokens = [1, 2, 3, 4, 5, 6, 7, 8, 9]
# jy: 模型 pretrain 中的多轮 epoch
for epoch in range(3):
    cand_indexes = []
    for i in input_tokens:
        cand_indexes.append(i)

    print("-" * 66 + str(epoch))
    # jy: cand_indexes 的值每次总是一致
    print("cand_indexes: %s" % cand_indexes)
    # jy: 基于 MLM 进行预训练时, 生成 mask 数组时会调用 transformers 包中的
    #      _whole_word_mask 方法, 而该方法中会基于 random.shuffle 进行随机
    #     mask; 而多轮 epoch 时如果是相同的句子, 则 cand_indexes 的值也是相
    #     同的, 不过尽管设置了 random.seed(42), 其保证的是每次运行该脚本也是
    #     得到相同的结果, 但脚本一次运行、多次对同一个 cand_indexes 进行 shuffle
    #     时得到的结果并不是一样的(因此对同一个句子进行多轮 epoch mask 操作
    #     时, 还是会尽可能的随机 mask 到不同的位置);
    random.shuffle(cand_indexes)
    print(cand_indexes)
    print("-" * 66 + str(epoch))
"""
------------------------------------------------------------------0
cand_indexes: [1, 2, 3, 4, 5, 6, 7, 8, 9]
[4, 7, 8, 5, 9, 3, 6, 1, 2]
------------------------------------------------------------------0
------------------------------------------------------------------1
cand_indexes: [1, 2, 3, 4, 5, 6, 7, 8, 9]
[8, 3, 6, 7, 1, 4, 5, 2, 9]
------------------------------------------------------------------1
------------------------------------------------------------------2
cand_indexes: [1, 2, 3, 4, 5, 6, 7, 8, 9]
[3, 6, 8, 9, 7, 2, 5, 1, 4]
------------------------------------------------------------------2
"""
`

**Comment by staoxiao** [comment]
> Actually, while you're here, a question about DupMAE. It looks like the modeling_duplex file you included doesn't have the actual pooling operation you describe in the paper - how is the dimension reduction from (768, 768) performed? Is it something like a simple PCA or is it a trainable layer in itself?
> 
> Also, what do you think the impact would be of a concatenated vector for the embeddings in terms of the continuity of the vector space along different axes? Would the space be less interpretable because (if I'm following) the CLS token is "meant" to store semantic information while the aggregated OT embeddings store fine-grained word information, and they're being concatenated into one?

For the OT embedding (30522 dimensions), we only store the top-768 value and its indices for retrieval. You can refer to splade repo: https://github.com/naver/splade

**Comment by mtybadger** [comment]
Thank you - the rest i can get from the paper. Is there a reason you didn’t train BGE with DupMAE instead of RetroMAE?

> > Actually, while you're here, a question about DupMAE. It looks like the modeling_duplex file you included doesn't have the actual pooling operation you describe in the paper - how is the dimension reduction from (768, 768) performed? Is it something like a simple PCA or is it a trainable layer in itself?
> > Also, what do you think the impact would be of a concatenated vector for the embeddings in terms of the continuity of the vector space along different axes? Would the space be less interpretable because (if I'm following) the CLS token is "meant" to store semantic information while the aggregated OT embeddings store fine-grained word information, and they're being concatenated into one?
> 
> For the OT embedding (30522 dimensions), we only store the top-768 value and its indices for retrieval. You can refer to splade repo: https://github.com/naver/splade

**Comment by staoxiao** [comment]
> Thank you - the rest i can get from the paper. Is there a reason you didn’t train BGE with DupMAE instead of RetroMAE?
> 
> > > Actually, while you're here, a question about DupMAE. It looks like the modeling_duplex file you included doesn't have the actual pooling operation you describe in the paper - how is the dimension reduction from (768, 768) performed? Is it something like a simple PCA or is it a trainable layer in itself?
> > > Also, what do you think the impact would be of a concatenated vector for the embeddings in terms of the continuity of the vector space along different axes? Would the space be less interpretable because (if I'm following) the CLS token is "meant" to store semantic information while the aggregated OT embeddings store fine-grained word information, and they're being concatenated into one?
> > 
> > 
> > For the OT embedding (30522 dimensions), we only store the top-768 value and its indices for retrieval. You can refer to splade repo: https://github.com/naver/splade

hi, the difference between retromae and dupmae is that the dupmae enhance the ability of token embeddings, which is used in sparse vector. In BGE project, we focus on the dense vector, so we only use retromae to improve the embedidng of cls.
